src/i256.rs

```diff
@@ -0,0 +1,279 @@
+use std::cmp;
+use std::convert::{TryFrom, TryInto};
+use std::ops::{Div, Mul, Neg, Sub};
+
+use uint::construct_uint;
+
+use crate::ArithmeticError;
+
+const TOTAL_BITS_COUNT: usize = 256;
+const UINT_WORD_BITS_COUNT: usize = 64;
+const UINT_WORDS_COUNT: usize = TOTAL_BITS_COUNT / UINT_WORD_BITS_COUNT;
+const SIGN_MASK: u64 = 1 << (UINT_WORD_BITS_COUNT - 1); // MSB = 1, other are equal to 0.
+
+// Single word has 64 bits. For 256-bit number:
+// UInt words count = 256 / 64 = 4
+construct_uint! {
+    pub struct U256(4);
+}
+
+/// Signed 256-bit number. Works on top of U256 with help of two's complement.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct I256 {
+    inner: U256,
+}
+
+impl I256 {
+    /// Value `i128::MAX`
+    pub const I128_MAX: Self = Self::from_i128(i128::MAX);
+    /// Value `i128::MIN`. Very useful because `abs` isn't defined for `i128::MIN`
+    pub const I128_MIN: Self = Self::from_i128(i128::MIN);
+    pub const MAX: Self = Self::new(U256([u64::MAX, u64::MAX, u64::MAX, !SIGN_MASK]));
+    pub const MIN: Self = Self::new(U256([0, 0, 0, SIGN_MASK]));
+
+    const fn new(x: U256) -> Self {
+        I256 { inner: x }
+    }
+
+    pub const fn from_i128(x: i128) -> Self {
+        let msb = if x < 0 { u64::MAX } else { 0 };
+        Self::new(U256([
+            x as u64,
+            (x >> UINT_WORD_BITS_COUNT) as u64,
```

Comment:
I don't see the point of a constant here because around a lot of `u64` and `u256`, and it's obvious what `>> 64` means here.

Reply:
done